spark-cassandra-connector/src/test/scala/org/apache/spark/sql/cassandra/PredicatePushDownSpec.scala

```diff
@@ -0,0 +1,214 @@
+  }
+
+  val pk1 = ColumnDef("pk1", PartitionKeyColumn, IntType)
+  val pk2 = ColumnDef("pk2", PartitionKeyColumn, IntType)
+  val c1 = ColumnDef("c1", ClusteringColumn(0), IntType)
+  val c2 = ColumnDef("c2", ClusteringColumn(1), IntType)
+  val c3 = ColumnDef("c3", ClusteringColumn(2), IntType)
+  val i1 = ColumnDef("i1", RegularColumn, IntType, indexed = true)
+  val i2 = ColumnDef("i2", RegularColumn, IntType, indexed = true)
+  val r1 = ColumnDef("r1", RegularColumn, IntType)
+  val r2 = ColumnDef("r2", RegularColumn, IntType)
+
+  val table = TableDef(
+    keyspaceName = "test",
+    tableName = "test",
+    partitionKey = Seq(pk1, pk2),
+    clusteringColumns = Seq(c1, c2, c3),
+    regularColumns = Seq(i1, i2, r1, r2)
+  )
+
+  "PredicatePushDown" should "push down an equality partition key predicate" in {
+    val f1 = EqFilter("pk1")
+    val f2 = EqFilter("pk2")
+    val ppd = new PredicatePushDown(Set[Filter](f1, f2), table)
+    ppd.predicatesToPushDown should contain allOf(f1, f2)
+    ppd.predicatesToPreserve shouldBe empty
+  }
+
+  it should "not push down a partition key predicate for a part of the partition key (1)" in {
+    val f1 = EqFilter("pk1")
+    val ppd = new PredicatePushDown(Set[Filter](f1), table)
+    ppd.predicatesToPushDown shouldBe empty
+    ppd.predicatesToPreserve should contain only f1
+  }
+
+  it should "not push down a partition key predicate for a part of the partition key (2)" in {
+    val f1 = EqFilter("pk2")
+    val ppd = new PredicatePushDown(Set[Filter](f1), table)
+    ppd.predicatesToPushDown shouldBe empty
+    ppd.predicatesToPreserve should contain only f1
+  }
+
+  it should "not push down a range partition key predicate" in {
+    val f1 = EqFilter("pk1")
+    val f2 = RangeFilter("pk2")
+    val ppd = new PredicatePushDown(Set[Filter](f1, f2), table)
+    ppd.predicatesToPushDown shouldBe empty
+    ppd.predicatesToPreserve should contain allOf(f1, f2)
+  }
+
+  it should "push down an IN partition key predicate on the last partition key column" in {
+    val f1 = EqFilter("pk1")
+    val f2 = InFilter("pk2")
+    val ppd = new PredicatePushDown(Set[Filter](f1, f2), table)
+    ppd.predicatesToPushDown should contain allOf(f1, f2)
+    ppd.predicatesToPreserve shouldBe empty
+  }
+
+  it should "not push down an IN partition key predicate on the non-last partition key column" in {
+    val f1 = InFilter("pk1")
+    val f2 = EqFilter("pk2")
+    val ppd = new PredicatePushDown(Set[Filter](f1, f2), table)
+    ppd.predicatesToPushDown shouldBe empty
+    ppd.predicatesToPreserve should contain allOf(f1, f2)
+  }
+
+  it should "push down the first clustering column predicate" in {
+    val f1 = EqFilter("c1")
+    val ppd = new PredicatePushDown(Set[Filter](f1), table)
+    ppd.predicatesToPushDown should contain only f1
+    ppd.predicatesToPreserve shouldBe empty
+  }
+
+  it should "push down the first and the second clustering column predicate" in {
```

Comment:
Functionally they are different and cover different usecases. It checks if a range predicate gets pushed down correctly. It is possible for an incorrect code to pass the previous test but fail this one. I can also imagine code failing the previous one and passing this one; if for example someone hardcoded a restriction that the last predicate must be a range.